src/toil/server/cli/wes_cwl_runner.py

```diff
@@ -0,0 +1,342 @@
+            "workflow_type_version": workflow_type_version
+        }
+
+        # Convert engine arguments into a JSON object
+        if workflow_engine_parameters:
+            params = {}
+            for param in workflow_engine_parameters:
+                if '=' not in param:  # flags like "--logDebug"
+                    k, v = param, None
+                else:
+                    k, v = param.split('=', 1)
+                params[k] = v
+            data["workflow_engine_parameters"] = json.dumps(params)
+
+        # Deal with workflow attachments
+        if attachments is None:
+            attachments = []
+        base = os.path.dirname(wf_url)
+        attachments.append(wf_url)
+
+        workflow_attachments = []
+        for file in attachments:
+            with open(file, "rb") as f:
+                rel: str = os.path.relpath(file, base)
+                if '../' in rel:
+                    # when inputs are in a different directory from the workflow
+                    rel = os.path.basename(file)
+                workflow_attachments.append((rel, BytesIO(f.read())))
+
+        return data, [("workflow_attachment", val) for val in workflow_attachments]
+
+    def run_with_engine_options(
+            self,
+            workflow_file: str,
+            secondary_file: Optional[str],
+            attachments: Optional[List[str]],
+            engine_options: Optional[List[str]]
+    ) -> Dict[str, Any]:
+        """
+        Composes and sends a post request that signals the WES server to run a
+        workflow.
+
+        :param workflow_file: A local/http/https path to a CWL/WDL/python
+                              workflow file.
+        :param secondary_file: A local path to a json or yaml file.
+        :param attachments: A list of local paths to files that will be
+                            uploaded to the server.
+        :param engine_options: A list of engine options to attach.
+
+        :return: The body of the post result as a dictionary.
+        """
+        data, files = WESClientWithWorkflowEngineParameters.build_wes_request(
+            workflow_file, secondary_file, attachments, engine_options)
+        post_result = requests.post(
+            f"{self.proto}://{self.host}/ga4gh/wes/v1/runs",
+            data=data,
+            files=files,  # type: ignore[arg-type]
+            auth=self.auth,
+        )
+
+        return cast(Dict[str, Any], wes_response(post_result))
+
+
+def get_deps_from_cwltool(cwl_file: str, input_file: Optional[str] = None) -> List[str]:
+    """
+    Return a list of dependencies of the given workflow from cwltool.
+
+    :param cwl_file: The CWL file.
+    :param input_file: Omit to get the dependencies from the CWL file. If set,
+                       this returns the dependencies from the input file.
+    """
+
+    option = '--print-input-deps' if input_file else '--print-deps'
+
+    args = ['cwltool', option, '--relative-deps', 'cwd', cwl_file]
```

Comment:
It looks like `printdeps` in the cwltool code eventually calls [`find_deps`](https://github.com/common-workflow-language/cwltool/blob/898fddf9167620645177eb94c8ea68ede0667f96/cwltool/main.py#L598) to actually get the deps. I would recommend importing and calling that, rather than running the command and parsing the output, unless replicating the setup for it would be hard.

Reply:
Here is a script that I'm using to find the dependencies for the CWL file, and it seems to work:

```python
from cwltool.context import LoadingContext
from cwltool.load_tool import resolve_tool_uri
from cwltool.main import find_deps
from schema_salad.exceptions import ValidationException

TOOL = "file:///Users/wlgao/code/GI/toil/test_files/hello.cwl"


try:
    loading_context = LoadingContext()
    uri, tool_file_uri = resolve_tool_uri(
        TOOL,
        resolver=loading_context.resolver,
        fetcher_constructor=loading_context.fetcher_constructor,
    )

    deps = find_deps(None, loading_context.loader, uri)
    print(deps)

except ValidationException:
    # handle error?
    raise
```

But there will be a lot more things to set up and load for `print_input_deps` to work, and we would need to duplicate a lot of code from `cwltool`: https://github.com/common-workflow-language/cwltool/blob/898fddf9167620645177eb94c8ea68ede0667f96/cwltool/main.py#L1287-L1322

Reply:
Maybe calling it on the command line is the best way then.